processing/src/main/java/io/druid/segment/incremental/OffHeapIncrementalIndex.java

```diff
@@ -0,0 +1,528 @@
+        } else {
+          throw new ISE("WTF! we are in sychronized block.");
+        }
+      }
+    }
+
+    rowContainer.set(row);
+
+    for (int i = 0; i < metrics.length; i++) {
+      final BufferAggregator agg = getAggs()[i];
+
+      synchronized (agg) {
+        agg.aggregate(aggBuffer, bufferOffset + aggOffsetInBuffer[i]);
+      }
+    }
+    rowContainer.set(null);
+    return numEntries.get();
+  }
+
+  @Override
+  public boolean canAppendRow()
+  {
+    final boolean canAdd = size() < maxRowCount;
+    if (!canAdd) {
+      outOfRowsReason = String.format("Maximum number of rows [%d] reached", maxRowCount);
+    }
+    return canAdd;
+  }
+
+  @Override
+  public String getOutOfRowsReason()
+  {
+    return outOfRowsReason;
+  }
+
+  @Override
+  protected BufferAggregator[] getAggsForRow(int rowOffset)
+  {
+    return getAggs();
+  }
+
+  @Override
+  protected Object getAggVal(BufferAggregator agg, int rowOffset, int aggPosition)
+  {
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.get(bb, indexAndOffset[1] + aggOffsetInBuffer[aggPosition]);
+  }
+
+  @Override
+  public float getMetricFloatValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.getFloat(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  public long getMetricLongValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.getLong(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  public Object getMetricObjectValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.get(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  /**
+   * NOTE: This is NOT thread-safe with add... so make sure all the adding is DONE before closing
+   */
+  @Override
+  public void close()
+  {
+    super.close();
+    facts.clear();
+    indexAndOffsets.clear();
+
+    if (selectors != null) {
+      selectors.clear();
+    }
+
+    RuntimeException ex = null;
+    for (ResourceHolder<ByteBuffer> buffHolder : aggBuffers) {
+      try {
+        buffHolder.close();
+      } catch(IOException ioe) {
+        if (ex == null) {
+          ex = Throwables.propagate(ioe);
+        } else {
+          ex.addSuppressed(ioe);
+        }
+      }
+    }
+    aggBuffers.clear();
+    if (ex != null) {
+      throw ex;
+    }
+  }
+
+  private static class OnHeapDimDim implements DimDim
```

Comment:
will do.
I started with copy paste mostly because that class was being changed by many ppl and I wanted to avoid merge conflicts on each fetch and rebase.